File: ant/src/com/appsecinc/ant/license/LicenseManager.java
package com.appsecinc.ant.license;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class LicenseManager {
	private static List<LicenseFile> _licenseFiles = new ArrayList<LicenseFile>();
	
	public LicenseManager() {
		_licenseFiles.add(new LICENSE_LicenseFile());
		_licenseFiles.add(new CPL_LicenseFile());
		_licenseFiles.add(new CPLv1_LicenseFile());
		_licenseFiles.add(new LICENSEHTM_LicenseFile());
		_licenseFiles.add(new COPYING_LicenseFile());
	}
	
	/**
	 * Find a license file in a directory.
	 * @param dir
	 *  Target directory.
	 * @return
	 *  License file.
	 */
	public List<LicenseFound> find(String root, String node, String product, String version, File dir, int depth) {
		List<LicenseFound> licenses = new ArrayList<LicenseFound>();
		String[] files = dir.list();
		for(String file : files) {
			
			File sub = new File(dir, file);
			if (sub.isDirectory())
				continue;

			for(LicenseFile licenseFile : _licenseFiles) {
				if (licenseFile.isMatch(file)) {					
					String subProduct = licenseFile.getSpec(file);
					if (subProduct != null && subProduct.length() > 0) {
						licenses.add(new LicenseFound(root, node, subProduct, product, null, licenseFile, sub));
					} else {					
						licenses.add(new LicenseFound(root, node, product, null, version, licenseFile, sub));
					}
				}
			}
		}
		
		return licenses.size() > 0 ? licenses : null;
	}
}

File: msbuild/src/AppSecInc.LicensesCollector/LicenseManager.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace AppSecInc.LicensesCollector
{
  class LicenseManager
  {
    private static List<ILicenseFile> _licenseFiles = new List<ILicenseFile>();

    public LicenseManager()
    {
      _licenseFiles.Add(new LICENSE_LicenseFile());
      _licenseFiles.Add(new CPL_LicenseFile());
      _licenseFiles.Add(new CPLv1_LicenseFile());
      _licenseFiles.Add(new LICENSEHTM_LicenseFile());
      _licenseFiles.Add(new COPYING_LicenseFile());
      _licenseFiles.Add(new GFDL_LicenceFile());
      _licenseFiles.Add(new LGPL_LicenseFile());
      _licenseFiles.Add(new IPL_LicenseFile());
      _licenseFiles.Add(new IDPL_LicenseFile());
    }

    /// <summary>
    /// Find a license file in a directory.
    /// </summary>
    /// <param name="dir">Target directory</param>
    /// <returns>License file</returns>
    public List<LicenseFound> Find(String root, String node, String product, String version, String dir, int depth)
    {
      List<LicenseFound> licenses = new List<LicenseFound>();

      DirectoryInfo directory = new DirectoryInfo(dir);
      FileInfo[] files = directory.GetFiles();
      foreach (FileInfo file in files)
      {
        if (isFolder(file.FullName))
          continue;

        foreach (ILicenseFile licenseFile in _licenseFiles)
        {
          if (licenseFile.IsMatch(file.Name))
          {
            String subProduct = licenseFile.GetSpec(file.Name);
            if (!String.IsNullOrEmpty(subProduct))
            {
              licenses.Add(new LicenseFound(root, node, subProduct, product, null, licenseFile, file.FullName));
            }
            else
            {
              licenses.Add(new LicenseFound(root, node, product, null, version, licenseFile, file.FullName));
            }
          }
        }
      }

      return licenses.Count > 0 ? licenses : null;
    }

    private bool isFolder(string path)
    {
      return ((File.GetAttributes(path) & FileAttributes.Directory) == FileAttributes.Directory);
    }
  }
}
